src/kitovu/sync/filecache.py

```diff
@@ -0,0 +1,124 @@
+"""
+
+import attr
+from enum import Enum
+import json
+import pathlib
+import typing
+
+from kitovu.sync import syncplugin
+
+
+class FileState(Enum):
+    """Used to discern in which places files have changed."""
+
+    NEW = 3
+    REMOTE_CHANGED = 4
+    NO_CHANGES = 5
+    LOCAL_CHANGED = 6
+    BOTH_CHANGED = 7
+
+
+@attr.s
+class File:
+
+    cached_digest: str = attr.ib()  # local digest at synctime
+    plugin_name: str = attr.ib()
+
+    def to_dict(self) -> typing.Dict[str, str]:
+        return {"plugin": self.plugin_name,
+                "digest": self.cached_digest}
+
+
+class FileCache:
+
+    def __init__(self, filename: pathlib.Path):
+        self._filename: pathlib.Path = filename
+        self._data: typing.Dict[pathlib.Path, File] = {}
+
+    def _compare_digests(self, remote_digest: str, local_digest: str, cached_digest: str) -> FileState:
+        local_changed: bool = local_digest != cached_digest
+        remote_changed: bool = remote_digest != cached_digest
+        if not remote_changed and not local_changed:  # case 5 above
+            return FileState.NO_CHANGES
+        elif remote_changed and not local_changed:  # case 4 above
+            return FileState.REMOTE_CHANGED
+        elif not remote_changed and local_changed:  # case 6 above
+            return FileState.LOCAL_CHANGED
+        elif remote_changed and local_changed:  # case 7 above
+            return FileState.BOTH_CHANGED
+
+    def write(self) -> None:
+        """"Writes the data-dict to JSON."""
+        json_data: typing.Dict[str, typing.Dict[str, str]] = {}
+
+        for key, value in self._data.items():
+            json_data[str(key)] = value.to_dict()
+
+        with self._filename.open("w") as f:
+            json.dump(json_data, f)
+
+    def load(self) -> None:
+        """This is called first when the synchronisation process is started."""
+        with self._filename.open("r") as f:
+            json_data = json.load(f)
+        for key, value in json_data.items():
+            digest: str = value["digest"]
+            plugin_name: str = value["plugin"]
+            self._data[pathlib.Path(key)] = File(cached_digest=digest, plugin_name=plugin_name)
+
+    def modify(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin, local_digest_at_synctime: str):
+        file = File(cached_digest=local_digest_at_synctime, plugin_name=plugin.NAME)
+        self._data[path] = file
+
+    def discover_changes(self,
+                         local_full_path: pathlib.Path,
+                         remote_full_path: pathlib.PurePath,
+                         plugin: syncplugin.AbstractSyncPlugin) -> FileState:
+        """Check if the file that is currently downloaded (path-argument) has changed.
+
+        Change is discovered between local file cache and local file."""
+        if not local_full_path.exists():
+            return FileState.NEW
+
+        file: File = self._data[local_full_path]
```

Comment:
We should probably handle the case where the local file exists (because the user downloaded it manually or whatever) but it's not in the cache. Not sure how to handle that, though.